src/typechecker.rs

```diff
@@ -0,0 +1,45 @@
+//! The [`TypeCheck`] trait enables a value to get resolved to a type.
+//! FIXME: MOAR DOC
+
+use crate::{Context, instruction::TypeDec};
+
+/// In order to avoid recomputing various types multiple times, instruction types should
+/// keep a [`CheckedType`] instance as a member function, which should be set and fetched
+/// using [`set_type()`] and [`get_type`]
+#[derive(Clone)]
+pub enum CheckedType {
+    Resolved(TypeDec),
+    Void,
+    Unknown,
+}
+
+impl Default for CheckedType {
+    fn default() -> CheckedType {
+        CheckedType::Unknown
+    }
+}
```

Comment:
Considering Void and Unknown are both possible, I think it might be dangerous to assume one over the other for Default, but that might just be me.

Reply:
`Void` needs to be used for statements, so it's not a default. It's very much something that exists and is not unknown